src/utils/getUserRewards.ts

```diff
@@ -0,0 +1,205 @@
+import { BigNumber, BigNumberish, providers } from "ethers";
+import axios from "axios";
+import { maxBN, minBN, now, WAD } from "../helpers";
+import { GraphUserBalances, UserBalance, formatGraphBalances } from "./graph";
+import { Market } from "./graph/getGraphMarkets/markets.types";
+import { getEpochsBetweenTimestamps, timestampToEpoch } from "./timestampToEpoch";
+import { RewardsDistributor__factory } from "@morpho-labs/morpho-ethers-contract";
+import addresses from "@morpho-labs/morpho-ethers-contract/lib/addresses";
+import { ages } from "../ages";
+export const getUserRewards = async (
+  address: string,
+  blockNumber?: number,
+  provider: providers.Provider = new providers.InfuraProvider(1)
+) => {
+  let timestampEnd = now();
+  if (blockNumber) {
+    const block = await provider.getBlock(blockNumber);
+    timestampEnd = block.timestamp;
+  }
+  const userBalances = await getUserBalances(
+    ages[0].subgraphUrl, // TODO: export the subgraphUrl
+    address.toLowerCase(),
+    blockNumber
+  );
+  const currentRewards = userBalancesToUnclaimedTokens(address, userBalances?.balances || [], timestampEnd);
+  const claimableRaw = require("../../distribution/merkleTree/currentDistribution.json").proofs[address.toLowerCase()];
+  const claimable = claimableRaw ? BigNumber.from(claimableRaw.amount) : BigNumber.from(0);
+  const currentEpochRewards = currentRewards.sub(claimable);
+
+  const currentEpoch = timestampToEpoch(timestampEnd);
+  let currentEpochProjectedRewards = currentRewards;
+  if (currentEpoch?.epoch.finalTimestamp)
+    currentEpochProjectedRewards = userBalancesToUnclaimedTokens(
+      address,
+      userBalances?.balances || [],
+      currentEpoch.epoch.finalTimestamp
+    ).sub(claimable);
+
+  let claimed = BigNumber.from(0);
+  let claimData = {};
+  if (claimable.gt(0)) {
+    const rewardsDisributor = RewardsDistributor__factory.connect(addresses.morphoDao.rewardsDistributor, provider);
+    claimed = await rewardsDisributor.claimed(address);
+    claimData = claimable.sub(claimed).gt(0)
+      ? {
+          rewardsDistributor: rewardsDisributor.address,
+          functionSignature: "claim(address,uint256,bytes32[])",
+          args: {
+            address,
+            amount: claimableRaw.amount,
```

Comment:
because we have to call the claim function with the total accumulated amount. the rewards distributor subtracts the claimed amount. This manipulation prevents users to claim twice.

Reply:
Okay I get it!